adapters/adhese/adhese.go

```diff
@@ -19,267 +13,205 @@ import (
 	}
 
-	targetKeys := make([]string, 0, len(targetParsed))
-	for key := range targetParsed {
-		targetKeys = append(targetKeys, key)
+	// Check if bidderExt.Bidder is empty before unmarshalling
+	if len(bidderExt.Bidder) == 0 {
+		return nil, []error{&errortypes.BadInput{Message: "bidderExt.Bidder is empty"}}
 	}
-	sort.Strings(targetKeys)
-
-	for _, targetKey := range targetKeys {
-		var targetingValues = targetParsed[targetKey].([]interface{})
-		parametersAsString += "/" + url.PathEscape(targetKey)
-		for _, targetRawValKey := range targetingValues {
-			var targetValueParsed = targetRawValKey.(string)
-			parametersAsString += targetValueParsed + ";"
-		}
-		parametersAsString = strings.TrimRight(parametersAsString, ";")
+	var params openrtb_ext.ExtImpAdhese
+	if err := jsonutil.Unmarshal(bidderExt.Bidder, &params); err != nil {
+		return nil, []error{&errortypes.BadInput{Message: fmt.Sprintf("Error unmarshalling bidder ext: %v", err)}}
 	}
 
-	return parametersAsString
-}
-
-func extractGdprParameter(request *openrtb2.BidRequest) string {
-	if request.User != nil {
-		var extUser openrtb_ext.ExtUser
-		if err := jsonutil.Unmarshal(request.User.Ext, &extUser); err == nil {
-			return "/xt" + extUser.Consent
-		}
+	if params.Account == "" || params.Location == "" {
+		return nil, []error{&errortypes.BadInput{
+			Message: "Missing required params: location, account",
+		}}
 	}
-	return ""
-}
 
-func extractRefererParameter(request *openrtb2.BidRequest) string {
-	if request.Site != nil && request.Site.Page != "" {
-		return "/xf" + url.QueryEscape(request.Site.Page)
+	// define a map of targets[] and pre-fill it with the slot
+	targets := extTargetsAdhese{
+		"SL": []string{makeSlot(params)},
 	}
-	return ""
-}
-
-func extractIfaParameter(request *openrtb2.BidRequest) string {
-	if request.Device != nil && request.Device.IFA != "" {
-		return "/xz" + url.QueryEscape(request.Device.IFA)
+	// add any additional targets to the map from the params
+	for k, v := range params.Targets {
+		targets[k] = v
 	}
-	return ""
-}
 
-func (a *AdheseAdapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
-	errs := make([]error, 0, len(request.Imp))
-
-	var err error
-
-	// If all the requests are invalid, Call to adaptor is skipped
-	if len(request.Imp) == 0 {
-		errs = append(errs, WrapReqError("Imp is empty"))
-		return nil, errs
+	// marshal the ext.adhese.bidder object into the ext field
+	modifiedExt, err := jsonutil.Marshal(extImpAdheseWrapper{
+		"adhese": targets,
+	})
+	if err != nil {
+		return nil, []error{&errortypes.BadInput{Message: fmt.Sprintf("Error marshalling modified ext: %v", err)}}
 	}
 
-	var imp = &request.Imp[0]
-	var bidderExt adapters.ExtImpBidder
+	// copy the request and dereference any pointers* and override the ext of the
+	modifiedRequest := *request
+	modifiedRequest.Imp[0].Ext = modifiedExt
```

Comment:
While you are only making a shallow copy here, PBS core does give your adapter a unique `Imp[]` array, so this is still safe. This does also mean that you will be sending the entire `Imp[]` array from the request, with only the first Imp's `ext` having been modified. I am not sure if that is what you want or not, so highlighting that here.